lecture_01/map/map.go

```diff
@@ -0,0 +1,201 @@
+package main
+
+import (
+	"fmt"
+	"math/rand"
+
+	helper "github.com/manav2401/core-dev-course/helpers"
+)
+
+func main() {
+	// initialize a string and print occurrences of each word in it
+	str := "core dev course hello dev world dev core"
+	printWordOccurrence(str)
+
+	// initialize a slice of integers
+	slice0 := generateRandomSlice(10)
+	printOccurringNumbers(slice0)
+
+	// declare a new slice
+	slice1 := generateRandomSlice(15)
+
+	// find overlapping numbers in slice0 and slice1
+	slice2 := findOverlappingNumbers(slice0, slice1)
+	fmt.Printf("Overlapping numbers in slice0: %v and slice1: %v are %v\n\n", slice0, slice1, slice2)
+
+	// find nth fibonacci number using memoization in maps
+	val1, m := fibonacci(5, make(map[int]int))
+	fmt.Printf("fibonacci(5): %d\n", val1)
+
+	val2, m := fibonacci(8, m)
+	fmt.Printf("fibonacci(8): %d\n", val2)
+
+	val3, _ := fibonacci(4, m)
+	fmt.Printf("fibonacci(4): %d\n\n", val3)
+
+	// create a random multi dimensional map and print the sorted version of it
+	// map[fee]map[nonce]exists
+	m2 := generateMultiDimensionalMap(3, 3)
+	printSortedMultiDimensionalMap(m2)
+}
+
+// generateRandomSlice fills random numbers in range 0-9 till length `l`
+func generateRandomSlice(l int) []int {
+	s := make([]int, l)
+	for i := 0; i < len(s); i++ {
+		s[i] = rand.Intn(10)
+	}
+
+	return s
+}
+
+// generateMultiDimensionalMap generates a nested map with random numbers
+func generateMultiDimensionalMap(l1, l2 int) map[int]map[int]struct{} {
+	m := make(map[int]map[int]struct{})
+
+	for i := 0; i < l1; i++ {
+		idx := rand.Intn(100)
+
+		for {
+			// check if the index already exists
+			if _, ok := m[idx]; ok {
+				idx = rand.Intn(100)
+			} else {
+				break
+			}
+		}
+
+		m[idx] = make(map[int]struct{})
+
+		for j := 0; j < l2; j++ {
+			// it's okay to override an existing sub-index
+			m[idx][rand.Intn(100)] = struct{}{}
+		}
+	}
+
+	return m
+}
+
+// printWordOccurrence prints occurrence of each word in a string
+func printWordOccurrence(str string) {
+	m := make(map[string]int, len(str))
+
+	s := []byte{}
+
+	for i := 0; i < len(str); i++ {
```

Comment:
Won't work with unicode strings.

Reply:
I see, didn't wanted to use the strings library so used this approach. Will update. Thanks!